components/script/dom/subtlecrypto.rs

```diff
@@ -0,0 +1,250 @@
+        algorithm: AlgorithmIdentifier,
+        extractable: bool,
+        key_usages: Vec<KeyUsage>,
+        comp: InRealm,
+    ) -> Rc<Promise> {
+        let normalized_algorithm = normalize_algorithm(cx, algorithm, "generateKey");
+        let promise = Promise::new_in_current_realm(comp);
+        if let Err(e) = normalized_algorithm {
+            promise.reject_error(e);
+            return promise;
+        }
+
+        let (task_source, canceller) = self
+            .global()
+            .as_window()
+            .task_manager()
+            .dom_manipulation_task_source_with_canceller();
+        let this = Trusted::new(self);
+        let trusted_promise = TrustedPromise::new(promise.clone());
+        let alg = normalized_algorithm.clone();
+        let _ = task_source.queue_with_canceller(
+            task!(generate_key: move || {
+                let subtle = this.root();
+                let promise = trusted_promise.root();
+                let key = match alg {
+                    Ok(NormalizedAlgorithm::AesKeyGenParams(key_gen_params)) => {
+                        subtle.generate_key_aes_cbc(key_usages, key_gen_params, extractable)
+                    },
+                    _ => Err(Error::NotSupported),
+                };
+                match key {
+                    Ok(key) => promise.resolve_native(&key),
+                    Err(e) => promise.reject_error(e),
+                }
+            }),
+            &canceller,
+        );
+
+        promise
+    }
+}
+
+const ALG_AES_CBC: &str = "AES-CBC";
+const ALG_AES_CTR: &str = "AES-CTR";
+const ALG_AES_GCM: &str = "AES-GCM";
+const ALG_AES_KW: &str = "AES-KW";
+const ALG_SHA1: &str = "SHA1";
+const ALG_SHA256: &str = "SHA256";
+const ALG_SHA384: &str = "SHA384";
+const ALG_SHA512: &str = "SHA512";
+const ALG_HMAC: &str = "HMAC";
+const ALG_HKDF: &str = "HKDF";
+const ALG_PBKDF2: &str = "PBKDF2";
+const ALG_RSASSA_PKCS1: &str = "RSASSA-PKCS1-v1_5";
+const ALG_RSA_OAEP: &str = "RSA-OAEP";
+const ALG_RSA_PSS: &str = "RSA-PSS";
+const ALG_ECDH: &str = "ECDH";
+const ALG_ECDSA: &str = "ECDSA";
+#[allow(dead_code)]
+static SUPPORTED_ALGORITHMS: &[&str] = &[
+    ALG_AES_CBC,
+    ALG_AES_CTR,
+    ALG_AES_GCM,
+    ALG_AES_KW,
+    ALG_SHA1,
+    ALG_SHA256,
+    ALG_SHA384,
+    ALG_SHA512,
+    ALG_HMAC,
+    ALG_HKDF,
+    ALG_PBKDF2,
+    ALG_RSASSA_PKCS1,
+    ALG_RSA_OAEP,
+    ALG_RSA_PSS,
+    ALG_ECDH,
+    ALG_ECDSA,
+];
+
+const NAMED_CURVE_P256: &str = "P-256";
+const NAMED_CURVE_P384: &str = "P-384";
+const NAMED_CURVE_P521: &str = "P-521";
+#[allow(dead_code)]
+static SUPPORTED_CURVES: &[&str] = &[NAMED_CURVE_P256, NAMED_CURVE_P384, NAMED_CURVE_P521];
+
+#[derive(Clone)]
+pub enum NormalizedAlgorithm {
+    #[allow(dead_code)]
+    Algorithm(Algorithm),
+    AesKeyGenParams(AesKeyGenParams),
+}
+
+#[allow(unsafe_code)]
+/// The spec states to run operation steps in parallel, but DOMString does not impl Send.
+unsafe impl Send for NormalizedAlgorithm {}
```

Comment:
This is one of those things that will bite us if we ever make use of the !Send properties of DOMString. I would prefer to explicitly convert the strings to a sendable type instead of using unsafe code here.

Reply:
Noted, I'll rework this then